Subject: [PATCH] skip schedule broadcast and startservice if the user is stopped

when a user is removed or stopped, the braodcast sent to the user
should be skipped and the same as starting service
Also both sendBroadcast and getContentProvider use UserController.
isUserRunningLocked instead of the method UserManager.exits

diff --git a/services/core/java/com/android/server/am/ActiveServices.java b/services/core/java/com/android/server/am/ActiveServices.java
index ae1aef6..ee4e5d3 100755
--- a/services/core/java/com/android/server/am/ActiveServices.java
+++ b/services/core/java/com/android/server/am/ActiveServices.java
@@ -343,7 +343,7 @@
 
         ServiceRecord r = res.record;
 
-        if (!mAm.mUserController.exists(r.userId)) {
+        if (!mAm.mUserController.isUserRunningLocked(r.userId, 0)) {
             Slog.w(TAG, "Trying to start service with non-existent user! " + r.userId);
             return null;
         }
diff --git a/services/core/java/com/android/server/am/BroadcastQueue.java b/services/core/java/com/android/server/am/BroadcastQueue.java
index ea901ce..fe2e2e2 100644
--- a/services/core/java/com/android/server/am/BroadcastQueue.java
+++ b/services/core/java/com/android/server/am/BroadcastQueue.java
@@ -1109,6 +1109,15 @@
                         + " to " + r.curApp + ": process crashing");
                 skip = true;
             }
+
+            if (!skip && !mService.mUserController.isUserRunningLocked(UserHandle.getUserId(
+                    info.activityInfo.applicationInfo.uid), 0)) {
+                Slog.w(TAG, "Skipping delivery to " + info.activityInfo.packageName + " / "
+                                + info.activityInfo.applicationInfo.uid
+                                + " : user is stopped");
+                skip = true;
+            }
+
             if (!skip) {
                 boolean isAvailable = false;
                 try {
